### DISCUSSION TITLE
Can custom claims be set per user?

### DISCUSSION THREAD

**Discussion by aksdb** [kanidm/kanidm#3219] [Q&A]
For example when integrating Nextcloud, it could come in handy to override the username (to allow mapping over already existing users) or the quota. I can only find documentation how to set claims on groups ... which would not be very fitting to manage quotas or username mappings. Am I missing something or is this currently not possible? Could it be possible at all?

**Comment by Firstyear** [comment]
This currently isn't possible. The quotas are likely better served by group mappings. For username mappings, you likely need a way to template what the username should be, and templating into claims is a deep rabbit hole we don't want to dive into yet as there are potential security risks.

**Reply by aksdb** [reply]
Ah I feared as much.

I am even contemplating that "custom attributes" in general might be a cool thing. Kinda like an own schema for your accounts. Those could be mapped (via groups) to claims for OIDC and to attributes for LDAP. That way it would even be possible to cause a user-group membership to yield a new `objectClass` on LDAP, or any other attribute another service might need.

Would you find something like this nice-to-have or would this rather go against your vision for kanidm? (Asking, because in case I stay motivated, I might simply try to dive into the code in give it a shot. This might just be the right excuse to finally do something with rust ^^).

**Reply by Firstyear** [reply]
It's not that it's "against our vision". It's more that "we want to do it right". So if you want to get involved, the first step is an issue with the feature request, followed by a design document so we can help and work out "what it will look like" and discuss the trades, risks, etc.

**Reply by Blacks-Army** [reply]
Couldn't this be done by creating a group for the person, when the person is created with the exact same name, so you could map an custom claim as a user id for nextcloud?

**Comment by vnghia** [comment]
Hello, do the work for this start somewhere ? I want this feature too and would like to see if I contribute something for it ? Thank you!

On my first thought, we should probably go with a dynamic claim maps but not a templating one. Something like we have a list of which static attributes to extract from a user (like `username`) for example. What do you think @Firstyear ?

**Reply by Firstyear** [reply]
Generally a feature like this starts with a design document. So if we were going to start somewhere, that would be it.

The other important part here is how we represent data internally, which matters here around how we want to deal with templates and projection of attributes. I think listing out some use cases in a design is a start, and then we can help guide on what you'de need to do to implement that.

**Reply by vnghia** [reply]
@Firstyear I haven't found much usecase actually. For me, I need `username` but it can be set with `prefer_short_username`. If you still think this is a good feature. I could open an issue and setup some information for further discussing.

**Reply by Firstyear** [reply]
All good, we do know some use cases for it, but we'll get to it when we can. :)

**Reply by vnghia** [reply]
@Firstyear, I have a usecase for this now. Regarding https://github.com/kanidm/kanidm/issues/2993, I totally agree with you about why `kanidm` breaks the tradition and doesn't have `family_name` and `given_name` fields (People fuck up my name all the time :D). However some applications do not work without it so I want to have an ability to set custom attributes on users and use that inside the claim maps to avoid repeating the claim maps everytime.

**Reply by Firstyear** [reply]
Would it be better to indicate to those applications that they are broken for requesting those values? Since we don't store family/given name, we don't really have anything we could project cleanly into them ....

**Reply by vnghia** [reply]
That's why I am suggesting that we have custom attributes on each user, some kind of key-value map and then set the custom claim map to that value by a special syntax like `{key}`.

**Reply by Firstyear** [reply]
Yes, this is a bigger piece of work though, and it's on the todo list.

**Comment by aljoscha-s** [comment]
I have a question regarding to this. I've played a bit with custom claims and it's possible to create a custom claim for a user directly. But I can't see this claim it in the ID token.

I have a user `alice`  which is a member of `my_group`
 
Created a claim for groups `idm_all_persons`, `my_group`, and for `alice` directly

 `kanidm system oauth2 update-claim-map my_client my_claim idm_all_persons "Claim-idm"` 
 `kanidm system oauth2 update-claim-map my_client my_claim my_group "Claim-group"` 
 `kanidm system oauth2 update-claim-map my_client my_claim alice "Claim-alice"` 

Added supplemental scope

`kanidm system oauth2 update-sup-scope-map my_client my_group new_claim`
`kanidm system oauth2 update-sup-scope-map my_client alice new_claim`

Checked client

```
oauth2_rs_claim_map: new_claim:idm_all_persons@idm.example.com:;:"Claim-idm"
oauth2_rs_claim_map: new_claim:my_group@idm.example.com:;:"Claim-group"
oauth2_rs_claim_map: new_claim:alice@idm.example.com:;:"Claim-alice"

oauth2_rs_scope_map: my_group@idm.example.com: {"email", "openid", "profile"}
oauth2_rs_sup_scope_map: my_group@idm.example.com: {"my_claim"}
oauth2_rs_sup_scope_map: alice@idm.example.com: {"my_claim"}
```

Now, when I check the ID token, which I got from the Kanidm. I see `my_claim` array with values from `idm_all_persons` and `my_group` but not from `alice`.

```
  "my_claim": [
    "Claim-idm",
    "Claim-group"
  ]
```

So does creating a custom claim for a user directly is meaningless?

 `kanidm system oauth2 update-claim-map my_client my_claim alice "Claim-alice"`

**Reply by Firstyear** [reply]
From the book:

"""
Claim values are mapped based on membership to groups. When an account is a member of multiple groups that would receive the same claim, the values of these maps are merged.
"""

https://kanidm.github.io/kanidm/stable/integrations/oauth2/custom_claims.html

While you can add the user as a claim recipient, since they aren't a group the claims aren't added ( https://github.com/kanidm/kanidm/blob/master/server/lib/src/idm/oauth2.rs#L3190 )

Custom claims is still the better approach here rather than trying to add all these claim maps.

**Reply by aljoscha-s** [reply]
@Firstyear Thanks for the reply!

Though the last sentence confused me

> Custom claims is still the better approach here rather than trying to add all these claim maps.

As I've understood, the only possibility to add custom claims is via claim maps. Is there some alternative approach to add custom claims to an account in Kanidm?

**Reply by Firstyear** [reply]
You can not provide claims to individual accounts today - only to groups, which provide the claims to their members.

**Reply by aljoscha-s** [reply]
Thanks a lot for the confirmation!

**Reply by Firstyear** [reply]
There is an open issuie for it somewhere, and I want to work on it soon though.

**Comment by andreas-bulling** [comment]
I have a problem for which what you propose in this issue also seems to be the only solution - if I understand correctly:
https://github.com/kanidm/kanidm/discussions/4406#discussioncomment-17375351

I have existing Nextcloud users for whom I cannot (or rather don't want to) change the internal UIDs. To allow them to be authenticated through kanidm, and accounts be linked properly, I have to add their unique UIDs to their person profiles in kanidm (as a new attribute/claim). But this doesn't seem to be possible at the moment?!

**Reply by Firstyear** [reply]
As in a unique uuid for nextcloud's usage? So the user alreaidy has a nextcloud-uuid internally and you need that exposed somehow?

**Reply by andreas-bulling** [reply]
yes - the users already have a unique NC-internal uuid and I need a way to (manually) add this uuid to their kanidm profiles (and the authentication token) so that I can map and compare these uuid's during the authentication process.

**Reply by Firstyear** [reply]
Right this is something we'd like to do but it's a long way away as it relies on a lot of things.

**Reply by Blacks-Army** [reply]
I can tell you: this is definitely possible. I got it working before.

The only thing I remember is that you need to create a group for each user and map the UUID claim to a custom scope and give this scope Nextcloud so it maps the UUID correctly. I don’t remember the exact steps anymore because I no longer use Nextcloud, but I can confirm that it works.

Maybe try using AI that’s actually how I managed to get it working back then.

**Comment by andreas-bulling** [comment]
As discussed here (https://github.com/kanidm/kanidm/discussions/4406#discussioncomment-17410844) this is indeed the reason for my problem with Nextcloud. And what will, unfortunately, prevent me from fully migrating away from FreeIPA to kanidm. I cannot manually change the internal UUIDs of 50+ people in the Nextcloud database. This is a guarantee for trouble. That's a real pity :((
